Novnode flag

mambocab opened the pull request:
Probably not ready for merge yet, just looking to start a conversation. @ptnapoleon?

ptnapoleon commented:
Looking good so far? I'm interested in where you're putting the logic for running a matrix

knifewine commented:
Didn't dive very deep, but I like it.

mambocab commented:
Pushed a couple small changes to names. I think this actually is ready for merge, though we would need to change the `novnodes` job to do so. Are we good with that? We will also be changing CassCI jobs up to use the testing matrix soon, so we might want to get all the CassCI churn out of the way in one shot.

mambocab commented:
Oh yeah, and you can smoke test this with this file:

https://gist.github.com/mambocab/31e94af3cf649f41ade7

mambocab commented:
The history needs some serious grooming and apparently Travis is mad, but I've changed the API here -- instead of just letting `nose.main` grab `sys.argv`, we're passing it an argv programatically. I think this implies some changes to the configuration plugin, since configuration will happen in the args to `runner.py` and not to `nosetests`.

Thoughts, @knifewine? I've already discussed this privately with @ptnapoleon and he says it is Fine.

mambocab commented:
Example command:

```sh
CASSANDRA_VERSION=git:trunk ./runner.py --nose-options '--nocapture --collect-only' putget_test.py
```

knifewine commented:
if it's tested and working, I'm a +1

mambocab commented:
Ok, code is re-ready for review. This now goes beyond the original scope -- it doesn't just replace the environment variable for turning off vnodes, it also uses `runner.py` to generate a test matrix.

I recommend reviewers mess around with this file as a smoke test:

https://gist.github.com/mambocab/31e94af3cf649f41ade7

A good starting point might be

```sh
wget https://gist.githubusercontent.com/mambocab/31e94af3cf649f41ade7/raw/f6ce5cc97ed8d72da3bf32e7ca43b7020d251fcc/novnode_smoke_test.py

set -x

CASSANDRA_VERSION=git:trunk ./runner.py 2>&1 novnode_smoke_test.py --nose-option="--with-id --verbosity=3 --nocapture" --debug
CASSANDRA_VERSION=git:trunk ./runner.py 2>&1 novnode_smoke_test.py --nose-option="--with-id --verbosity=3 --nocapture" --debug --vnodes true
CASSANDRA_VERSION=git:trunk ./runner.py 2>&1 novnode_smoke_test.py --nose-option="--with-id --verbosity=3 --nocapture" --debug --vnodes false
CASSANDRA_VERSION=git:trunk ./runner.py 2>&1 novnode_smoke_test.py --nose-option="--with-id --verbosity=3 --nocapture" --debug --vnodes true --vnodes false
```

Do all of these behave as you expect?

mambocab commented:
(ping @ptnapoleon and @knifewine for review!)

ptnapoleon commented on plugins/dtestconfig.py:
```diff
@@ -0,0 +1,39 @@
+    enabled = True  # if this plugin is loaded at all, we're using it
+    name = 'dtest_config'
+
+    def __init__(self, config=None):
```
Could you use the doxygen `@param` markup in here?

Also, it doesn't seem that we are checking that `config` is a `GlobalConfigObject`

mambocab replied:
Just pushed the markup.

I don't think it's good style to check types. I think the idiomatic thing to do is just let it fail where it'd fail if it doesn't quack like a duck.

ptnapoleon commented on runner.py:
```diff
@@ -0,0 +1,205 @@
+
+cluster configuration options:
+    --vnodes VNODES_OPTIONS...   specify whether to run with or without vnodes.
+                                 valid values: 'true' and 'false'
```
nit: valid options

mambocab replied:
Based on the documentation for `docopt`, I think that `VNODES_OPTIONS` is an "argument" and valid "values" for it are `true` and `false`. That's my understanding.

ptnapoleon commented on runner.py:
```diff
@@ -0,0 +1,205 @@
+
+def _validate_and_serialize_vnodes(vnodes_value):
+    """
+    Validate the values recieved for vnodes configuration. Returns a
```
nit: `received`.

ptnapoleon commented on runner.py:
```diff
@@ -0,0 +1,205 @@
+
+    # function for printing debug statements
+    if options['--debug']:
+        debug = print
```
Does this work in py2?

mambocab replied:
Yes. I started the file with `from __future__ import print_function`.

ptnapoleon replied:
Good thing I noticed that.

ptnapoleon commented on runner.py:
```diff
@@ -0,0 +1,205 @@
+        temp = NamedTemporaryFile(dir=getcwd())
+        debug('Writing the following to {}:'.format(temp.name))
+        debug(to_execute)
+        debug()
```
Do we really need this? Can't you just add a newline on the previous call

mambocab replied:
Sure.

ptnapoleon commented on runner.py:
```diff
@@ -0,0 +1,205 @@
+        temp.flush()
+
+        results.append(subprocess.call(['python', temp.name]))
+        print()
```
Why print() here?

mambocab replied:
I'll add a comment. The `subprocess.call` call uses `stdout` and `stderr` from the parent process, and the `print()` call adds a newline after all that gets printed, separating it from the beginning of the next `subprocess.call` output.

ptnapoleon commented on runner.py:
```diff
@@ -0,0 +1,205 @@
+    """
+    Transforms a dictionary of {key: list(configuration_options} into a list of
+    dictionaries that takes the product of the lists at the keys and returns a
+    tuple of dictionaries, each corresponding to a point in the product, with
```
So which are we returning? A list of dicts, or a tuple of dicts?

mambocab replied:
Good catch, thanks.

mambocab commented:
@ptnapoleon I believe I've addressed all your comments in the most recent commit.

Can you think of a better way to explain what `product_of_values` does? because, as you can tell, I had a heck of a time.

ptnapoleon commented:
The new version of `product_of_values` is much better.

ptnapoleon commented:
I am +1

ptnapoleon commented:
Running `runner.py` without options should print out the help text, not

```
 python runner.py
Traceback (most recent call last):
  File "runner.py", line 30, in <module>
    nose_option_list = options['--nose-option'].split()
AttributeError: 'NoneType' object has no attribute 'split'
```

ptnapoleon commented:
Actually anytime we leave out nose options it throws that exception.

ptnapoleon commented:
Shouldn't `python runner.py --nose-option '-xsv' topology_test.py` run with and without vnodes?

ptnapoleon commented:
Why am I seeing this?
```
python runner.py --nose-option '-xsv' topology_test.py --debug
Usage: runner.py [--nose-option NOSE_OPTIONS] [TESTS...]
```

ptnapoleon commented:
Ugh, I've been running against the wrong SHA. H/o

ptnapoleon commented:
I'm passing `--nose-option` instead of `--nose-options`, and it isn't complaining. Also not working, at least.

ptnapoleon commented:
`--nose-options '-xs -vvv'` Is as verbose as I'd expect, but `--nose-options '-xsv'` definitely is not. It seems like something is wrong

mambocab commented:
> `--nose-options '-xs -vvv'` Is as verbose as I'd expect, but `--nose-options '-xsv'` definitely is not. It seems like something is wrong

See the comment here 

https://github.com/riptano/cassandra-dtest/commit/6963764ec7b26c5a2313bbf167f4938b30bfbe28#diff-2ef65ef8520cef01c37a957e90545296R208

:angry: Dangit, computers.

mambocab commented:
> I'm passing `--nose-option` instead of `--nose-options`, and it isn't complaining.

Turns out that is a feature, not a bug:

https://github.com/docopt/docopt/blob/master/docopt.py#L308

```
❯ cat wat.py
"""
Usage: wat.py [--options OPTIONS...]
"""
from docopt import docopt
from pprint import pprint

if __name__ == '__main__':
    pprint(docopt(__doc__))
                                                                                                                                  
❯ python wat.py --opti WAT
{'--options': True,
 'OPTIONS': ['WAT']}
```

`docopt` will assume that, e.g. `--opti` means `--option` because it's an unambiguous prefix. Same goes for `--nose-option` and `--nose-options`.

knifewine commented:
if possible, I'd love some output on what's about to happen before tests run. debug gives some output but I think that only covers the first setting permutation at the beginning.

ideally we'd be able to see up front all permutations before the first one starts off (if possible). rather than having this tied to --debug, which may include other verbose output, maybe we make it something like --preview/--no-preview ?

we should maybe formalize how --debug of runner.py relates to debug output in dtest -- are they the same or distinct?

mambocab commented:
> ideally we'd be able to see up front all permutations before the first one starts off (if possible). rather than having this tied to --debug, which may include other verbose output, maybe we make it something like --preview/--no-preview ?

I can just use a verbosity flag, like nose does -- `-v` is basically `quiet`, `-vv` is default, with the preview, `-vvv` is debug. Sound reasonable?

> we should maybe formalize how --debug of runner.py relates to debug output in dtest -- are they the same or distinct?

They're completely distinct; debug stuff in `dtest` is still (unfortunately) controlled by environment variables. The right way to fix that is probably to use built-in logging facilities, since `nose` plays nicely with that.

ptnapoleon commented:
I'd like a verbosity flag, yes. But as mentioned offline, that isn't quite nose's behavior. And yes the two debug outputs should be different

knifewine commented:
If some kind of permutation preview (all shown before running starts) is the default, I'm happy. In my mind this would be the first level of debug output, with other noise being opted-in to at higher levels.

knifewine commented:
I'm still a bit uneasy about runner.py and dtest both having something called "debug" because I think that's likely to be a point of confusion, but I also don't want to tangle up the command line arguments code. Seems that eventually this script could also have an argument for dtest debugging? -- so ideally we just give the runner debug cmdline a distinct name to leave room for the dtest debug option later?

ptnapoleon commented:
I agree with Russ.

mambocab commented:
Just pushed a couple more changes to improve debug output and change the names of the `--debug` and `--quiet` flags to `--runner-{debug,quiet}`.

knifewine commented:
one more user surface nit, I think 'runner.py' should maybe be renamed to something like 'testrunner.py' (well, crud, I guess that might mean nose would try and pick it up).

right now with a name like 'runner' it's kinda non-obvious that it's meant to be the test running entry point.

mambocab commented:
At @knifewine's suggestion, I added the output about the full config matrix to the default output. I also added a `--dry-run` option so users can see the effects of the options they specify before running the tests.

mambocab commented:
I've also renamed the runner. We don't have to worry about test discovery; `nosetests` automatically excludes executable files.

knifewine commented:
I got my console output wishes, i'm +1 based on my local testing (didn't dig into the code itself).

ptnapoleon commented:
+1

mambocab commented:
I've smushed all of these into one big commit. Now the question is: how do we want to handle CassCI jobs post-merge? The regular jobs -- the ones that run with vnodes -- should be unaffected. We will either need to 1) change all the novnodes jobs to use the runner script with `--vnodes false` or 2) migrate to matrix jobs. [Number 2 isn't ready yet](http://cassci.datastax.com/job/mambocab-sample-matrix-job/10/testReport/) but I'll keep working at that. So maybe number 1? Thoughts, @ptnapoleon and @mshuler?

mshuler commented:
No.1 should be a simple template edit.

As long as we work through all the details to making sure matrix jobs do most of the things we want, No.2 is fine.
* do they work with autojobs templates?
* does the jenkins reporting work for us?
* does that work with cstar_report, too?
* does the bucketizer work with matrices? (seems to be working in the test job)
* can we tell which report goes with which matrix option easily?

What do we gain by using matrix jobs? :)

mambocab commented:
> * do they work with autojobs templates?

I don't know enough about how templating works -- I'm guessing we'd have to make a new matrix template job.

The other option is to migrate to the [job creator](https://github.com/riptano/job-creator) for this. I have no idea how that'd interact with `autojobs`' templating.

> * does the jenkins reporting work for us?

Remains to be seen what the reporting will actually look like; I still have to fix these jobs up.

> * does that work with cstar_report, too?

Oof, great question, and one I hadn't looked at at all.

> * does the bucketizer work with matrices? (seems to be working in the test job)

Yep, not a problem as far as I can tell. Each subjob is more or less a normal CassCI job, just kicked off differently.

> * can we tell which report goes with which matrix option easily?

Same as above -- I don't know yet, we'll see how reporting turns out.

> What do we gain by using matrix jobs? :)

The gain now is that we'll be able to delete all the `*novnode_dtest` jobs:

http://cassci.datastax.com/search/?q=novnode_dtest

This is a maintenance win for us -- we wind up running the same tests under the same conditions, but with fewer jobs to keep updated and to monitor. In the long term, it'll also be pretty easy to add rows to the matrix for things like offheap memtables.

mambocab commented:
I have set up the jenkins template for novnodes jobs and am merging this.
